### ISSUE TITLE
Docker 1.8 corrupts /etc/hosts files in containers

### ISSUE THREAD

**Issue by thockin** [docker/docker#17190]
related: #16619

@mavenugo  - We have a problem.

We've been experiencing periodic end-to-end test failures where some containers can not resolve "localhost".  Killing the container and letting it restart usually fixes the problem.  Today we got it into this state and pinned down the following:

The container in question has a /etc/hosts file that is several hundred lines long.  That /etc/hosts file does NOT have localhost or the container's own IP.  If I add those 2 lines the tests pass.

Of course I thought of #16619 .  So we looked at /etc/hosts of every container on this machine - 84 in total.  Those files range in size from 175 lines (correct: 7 boilerplate + 2 per container) to 1700+ lines.  Looking at the offending files, there are a lot of duplicate IPs.  It looks like old containers have not been removed, but localhost, "self", and the ipv6 boilerplate has been removed.

Even looking at my local test machine (far less churn) I see the same.  I have 4 containers with network namespace running, I should have 8+7=15 lines, right?

```
# wc -l /var/lib/docker/containers/*/hosts
  15 /var/lib/docker/containers/3fc23782814513d1934ce21c0f1b096263df1fcacee6787208c3cda715822356/hosts
  15 /var/lib/docker/containers/4b7e8432a368eeb7f1dc49230f16c1d7c788b471e4dfa7d8fbf709c3495f430f/hosts
  19 /var/lib/docker/containers/5715aa09fc091ee0ff3d4dcf2ee300d8e52603dafcde8f24aaf1def960cf4a28/hosts
  15 /var/lib/docker/containers/6e7fe564dece829de3c335400357b2ef327a1a44c638045350f0b49e5324a189/hosts
  15 /var/lib/docker/containers/7845351acbacf0588142a1ebfa3fa481c3191da48a188b20c5d5fdae689cef1e/hosts
  17 /var/lib/docker/containers/ba23cbabfa90700f4d9ba80c5723afb1d7479a08b12efdded74e5874991ce76b/hosts
  17 /var/lib/docker/containers/c1e64038b35a0527b6aef911dad002278569724263cc922efd8e715747e83538/hosts
  19 /var/lib/docker/containers/fc7c5475fd923b75a0595431dcaf6b6db5fa8a18409d0df4e6282e3c0bd8aedf/hosts
```

Diffing a 15 and 19 line file shows:

```
# diff -u -U 100 <(sort -n /var/lib/docker/containers/3fc23782814513d1934ce21c0f1b096263df1fcacee6787208c3cda715822356/hosts) <(sort -n /var/lib/docker/containers/5715aa09fc091ee0ff3d4dcf2ee300d8e52603dafcde8f24aaf1def960cf4a28/hosts)
--- /dev/fd/63	2015-10-19 23:59:20.618879795 +0000
+++ /dev/fd/62	2015-10-19 23:59:20.618879795 +0000
@@ -1,15 +1,19 @@
 ::1	localhost ip6-localhost ip6-loopback
 fe00::0	ip6-localnet
 ff00::0	ip6-mcastprefix
 ff02::1	ip6-allnodes
 ff02::2	ip6-allrouters
+10.244.0.10	distracted_ritchie
+10.244.0.10	distracted_ritchie.bridge
 10.244.0.2	k8s_POD-7be6d81d_fluentd-cloud-logging-kubernetes-minion-isfq_kube-system_dcfde077753b8b6908952cfea6a14200_c4f00d1d
 10.244.0.2	k8s_POD-7be6d81d_fluentd-cloud-logging-kubernetes-minion-isfq_kube-system_dcfde077753b8b6908952cfea6a14200_c4f00d1d.bridge
 10.244.0.3	k8s_POD-7be6d81d_heapster-v10-lx8q7_kube-system_2a3fc6d6-6ddb-11e5-9249-42010af00002_cbe79e10
 10.244.0.3	k8s_POD-7be6d81d_heapster-v10-lx8q7_kube-system_2a3fc6d6-6ddb-11e5-9249-42010af00002_cbe79e10.bridge
 10.244.0.4	k8s_POD-c5371ceb_monitoring-influxdb-grafana-v2-koh3j_kube-system_2a43f315-6ddb-11e5-9249-42010af00002_76d3cb92
 10.244.0.4	k8s_POD-c5371ceb_monitoring-influxdb-grafana-v2-koh3j_kube-system_2a43f315-6ddb-11e5-9249-42010af00002_76d3cb92.bridge
 10.244.0.5	k8s_POD-9db2f941_kube-ui-v2-v4yd6_kube-system_2a42e6da-6ddb-11e5-9249-42010af00002_69a0a0c7
 10.244.0.5	k8s_POD-9db2f941_kube-ui-v2-v4yd6_kube-system_2a42e6da-6ddb-11e5-9249-42010af00002_69a0a0c7.bridge
-10.244.0.5	kube-ui-v2-v4yd6
+10.244.0.6	k8s_POD-6e934112_kube-dns-v9-ee38a_kube-system_2a4c2763-6ddb-11e5-9249-42010af00002_86e49fbc
+10.244.0.6	k8s_POD-6e934112_kube-dns-v9-ee38a_kube-system_2a4c2763-6ddb-11e5-9249-42010af00002_86e49fbc.bridge
+10.244.0.6	kube-dns-v9-ee38a
 127.0.0.1	localhost
```

Is there any reason the hosts files would be different for different containers on the same node?

I know docker is live-editing the hosts file (which is absolutely unsafe, but should only be unsafe for the users in the container).  Any clues on what's going on?

@artfulcoder

**Comment by phemmer** [comment]
Yeah, modification of this file is rather racy. I can corrupt the file very easily with the following:

```
docker run -d --name=test-master busybox sleep 99999999

for i in {00..09}; do docker create --name test-$i --link test-master busybox sleep inf; done

for i in {00..09}; do ( for (( j=0; j < 20; j++ )); do docker start test-$i; docker kill test-$i; done ) & done; wait

for i in {00..09}; do docker kill test-$i; done

docker exec test-master cat /etc/hosts
```

**Comment by brendandburns** [comment]
The basic problem is that Docker is using `ioutil.WriteFile` from golang:

https://github.com/docker/docker/blob/master/vendor/src/github.com/docker/libnetwork/etchosts/etchosts.go#L66
https://github.com/docker/docker/blob/master/vendor/src/github.com/docker/libnetwork/etchosts/etchosts.go#L93

which uses truncate and write:
https://golang.org/pkg/io/ioutil/#WriteFile

Truncate and write is inherently non-atomic and racy.

The only safe-ish thing to do here is to use file rename instead of truncate and write, this still isn't 100% atomic, but it is dramatically safer (eg. http://stackoverflow.com/questions/7054844/is-rename-atomic for details)

Honestly, the only truly safe thing to do here is to use an actual DNS server like we do in Kubernetes, instead of mucking with `/etc/hosts`

**Comment by thockin** [comment]
rename should be atomic, BUT it doesn't work across a bind mount.

**Comment by LK4D4** [comment]
I agree that DNS would be cool. Messing with hosts also slow (in addition to races) if you have a lot of containers.

**Comment by thockin** [comment]
All that aside - docker 1.8 seems to be totally broken in this regard.  Can
we just get this turned off for a 1.8.4?

**Comment by brendandburns** [comment]
yes please, second the request for a flag so that we can disable /etc/hosts writing.

**Comment by phemmer** [comment]
> The basic problem is that Docker is using ioutil.WriteFile from golang

I think it's more than that.
Docker is preserving custom entries in `/etc/hosts`. It doesn't rebuild it from scratch using only docker's own data. This means that you've got docker reading the file, making changes, then writing it back out. If you get 2 things doing this at the same time, one of them is going to discard the changes of the other.

**Comment by rade** [comment]
>  request for a flag so that we can disable /etc/hosts writing

You are not the first person to ask: https://github.com/docker/libnetwork/issues/519

In weave we introduced a [hack that prevents docker from updating `/etc/hosts`](https://github.com/weaveworks/weave/blob/5ff740cddc30ac87fd642116f990263a315a718d/weave#L1024). We are not proud of it, but it's the only option available to us.

**Comment by mrjana** [comment]
@thockin The first order of business is to fix the racy behavior of writing to /etc/hosts, which btw has existed since `--links` existed. Just that with CNM we are updating /etc/hosts of all containers in a given network. That makes live updating more common. 

Disabling /etc/hosts update is a UI discussion and has implications on the `--links` option and should be discussed separately as apart of #16619

**Comment by thockin** [comment]
@mrjana 2 thoughts.

1) I don't know of any way to update a file across a bind-mount atomically.  The usual `rename()` technique simply doesn't work.  I don't think this can be made safe.

```
thockin@freakshow:/tmp/test-bind-rename$ echo first > hosts

thockin@freakshow:/tmp/test-bind-rename$ echo nothing > container.hosts

thockin@freakshow:/tmp/test-bind-rename$ sudo mount --bind hosts container.hosts 

thockin@freakshow:/tmp/test-bind-rename$ cat hosts 
first

thockin@freakshow:/tmp/test-bind-rename$ cat container.hosts 
first

thockin@freakshow:/tmp/test-bind-rename$ echo second > container.hosts 

thockin@freakshow:/tmp/test-bind-rename$ cat hosts 
second

thockin@freakshow:/tmp/test-bind-rename$ cat container.hosts 
second

thockin@freakshow:/tmp/test-bind-rename$ echo third > new

thockin@freakshow:/tmp/test-bind-rename$ mv new hosts

thockin@freakshow:/tmp/test-bind-rename$ cat hosts 
third

thockin@freakshow:/tmp/test-bind-rename$ cat container.hosts 
second
```

Even hardlinks don't work.

```
thockin@freakshow:/tmp/test-bind-rename$ echo first > hosts

thockin@freakshow:/tmp/test-bind-rename$ ln hosts container.hosts

thockin@freakshow:/tmp/test-bind-rename$ cat hosts 
first

thockin@freakshow:/tmp/test-bind-rename$ cat container.hosts 
first

thockin@freakshow:/tmp/test-bind-rename$ echo second > container.hosts 

thockin@freakshow:/tmp/test-bind-rename$ cat hosts 
second

thockin@freakshow:/tmp/test-bind-rename$ cat container.hosts 
second

thockin@freakshow:/tmp/test-bind-rename$ echo third > new

thockin@freakshow:/tmp/test-bind-rename$ mv new hosts

thockin@freakshow:/tmp/test-bind-rename$ cat hosts 
third

thockin@freakshow:/tmp/test-bind-rename$ cat container.hosts 
second
```

2) At least with plain docker links you edit once at startup (I think), so it's far less dangerous (but maybe I am wrong).  Either way, it's a much reduced exposure than updating every container for every other container.  I think this needs an ASAP mititgation.

**Comment by mrjana** [comment]
@thockin 

For 1) We are going to experiment with acquiring a write lease(using `F_SETLEASE`) during the short period when live updating the /etc/hosts from docker. When we have write lease all other `opens` for writing and `truncate` will block for more than a few seconds(/proc/sys/fs/lease-break-time) and that is sufficient enough for us to complete the update. 

For 2) even with links it wasn't just during startup. everytime the child container restarted the parent container's hosts file gets live updated with the child container's new IP

Anyway, I do agree we need to mitigate this race quickly.

**Comment by thockin** [comment]
I did not know about SETLEASE - I learned something.  I also tried it, and
while F_SETLEASE seems to work on its own, it does NOT seem to work across
bind mounts.

```
thockin@freakshow:/tmp/test-bind-rename$ cat lease.c
#define _GNU_SOURCE 1
#include <unistd.h>
#include <fcntl.h>
#include <stdlib.h>
#include <signal.h>

int main(int argc, char *argv[]) {
    signal(SIGIO, SIG_IGN);
    int fd = open("hosts", O_RDWR);
    fcntl(fd, F_SETLEASE, F_WRLCK);
    sleep(10);
    exit(0);
}

thockin@freakshow:/tmp/test-bind-rename$ make lease
cc     lease.c   -o lease

thockin@freakshow:/tmp/test-bind-rename$ echo first > hosts

thockin@freakshow:/tmp/test-bind-rename$ echo nothng > container.hosts

thockin@freakshow:/tmp/test-bind-rename$ ./lease & time cat hosts
[1] 20737
first
[1]+  Done                    ./lease

real 0m10.001s
user 0m0.000s
sys 0m0.116s

thockin@freakshow:/tmp/test-bind-rename$ sudo mount --bind hosts
container.hosts

thockin@freakshow:/tmp/test-bind-rename$ cat hosts container.hosts
first
first

thockin@freakshow:/tmp/test-bind-rename$ ./lease & time cat hosts[1] 20953
first

real 0m0.001s
user 0m0.000s
sys 0m0.001s

thockin@freakshow:/tmp/test-bind-rename$ ./lease & time cat container.hosts
[2] 20977
first

real 0m0.008s
user 0m0.000s
sys 0m0.001s
```

Same behavior with hard links.  My guess is that Linux refuses to promise a
lease on dirents with a link count > 1.

**Comment by thockin** [comment]
http://lxr.free-electrons.com/source/fs/locks.c#L1576

```
1587         if ((arg == F_WRLCK) && ((d_count(dentry) > 1) ||
1588             (atomic_read(&inode->i_count) > 1)))
1589                 ret = -EAGAIN;
```

**Comment by thockin** [comment]
@mavenugo @mrjana @icecrime 

Options that I see:

1) Fix the merge logic to produce correct results, don't bind-mount /etc/hosts, use something like nsenter to do a proper atomic rename in each container's mount namespace, cut 1.8.4

2) Fix the merge logic to produce correct results, bind-mount /etc/hosts.$version, symlink /etc/hosts -> /etc/hosts.$version, use something like nsenter to do a proper atomic rename of a new symlink in each container's mount namespace, cut 1.8.4

3) Implement a flag to turn this whole function off, cut 1.8.4, fix it better in 1.9

#3 seems like the least risky, most likely to succeed plan.  We can help produce the patch, if needed, though it will be faster if you guys do it (knowing where to look in the code is 75% of the problem).

What do you guys think?

**Comment by rade** [comment]
@thockin I suspect 1 and 2 would break [weave's hack that prevents Docker from updating `/etc/hosts`](https://github.com/weaveworks/weave/blob/5ff740cddc30ac87fd642116f990263a315a718d/weave#L1024).

Option 3, please.

**Comment by thockin** [comment]
Option 3 is our preferred plan, too, and that's the patch we'll carry if we have to.

I know 1.9.x is at RC, but this bug nullifies everything we have done to qual 1.8.x so we're begging for a 1.8.4 patch.

**Comment by brendandburns** [comment]
I'm personally willing to develop the patch for 1.8.4, I've dug through the code a bit, and I'm pretty sure I know where to insert the right flag points.

We're likely to develop this patch one way or the other, and we'll just carry a patched 1.8.4 in Kubernetes if we have to, but we would obviously very much rather not.

**Comment by mrjana** [comment]
@thockin Getting back to fixing the racy /etc/hosts update here's something that might be the best possible solution:
```
#define _GNU_SOURCE 1
#include <unistd.h>
#include <fcntl.h>
#include <stdlib.h>
#include <signal.h>
#include <string.h>

int main(int argc, char *argv[]) {
  signal(SIGIO, SIG_IGN);
  int fd = open("hosts", O_RDWR);
  int rc;
  struct flock fl;

  memset(&fl, 0, sizeof fl);
  fl.l_type = F_WRLCK;
  fl.l_whence = SEEK_SET;
  fl.l_start = 0;
  fl.l_len = 0;

  rc = fcntl(fd, F_SETLKW, &fl);
  if (rc == -1) {
    perror("fcntl");
  }
  sleep(10);
  rc = fcntl(fd, F_UNLCK, &fl);
  if (rc == -1) {
    perror("fcntl unlock");
  }
  exit(0);
}
```

This point is to employ mandatory locking and confirmed with the following that it works across bind mounts:
```
mrjana@dev-1:/tmp/lease$ make lease
cc     lease.c   -o lease
mrjana@dev-1:/tmp/lease$
mrjana@dev-1:/tmp/lease$ echo first > hosts
mrjana@dev-1:/tmp/lease$ touch container.hosts
mrjana@dev-1:/tmp/lease$
mrjana@dev-1:/tmp/lease$ sudo mount -o bind,mand hosts container.hosts
mrjana@dev-1:/tmp/lease$ chmod g+s,g-x container.hosts
mrjana@dev-1:/tmp/lease$ ./lease & sleep 0.1 && time cat container.hosts
[1] 11287
first
[1]+  Done                    ./lease

real    0m10.777s
user    0m0.000s
sys     0m0.000s
```

Works across bind mounts(I acquire the lock in `hosts` while I try to read the bind mounted `container.hosts`)

But there is a catch. In some rare cases when the write system call is well past the check for lock and we try to acquire the lock, the lock will still be sucessfull. So it's not 100% race free but very close.

**Comment by thockin** [comment]
I just tried this.  It's clever.  PROBLEM: The host FS needs to be mounted
`-o mand` or the lock is just ignored.  You can't safely assume that every
FS is mounted with -o mand.  I'm not even sure every FS supports it.

Maybe you can mount all of /var/lib/docker/containers as a tmpfs with -o
mand?  That's a little sketchy.

That tiny bit of race is still pretty messy, but obviously better.  I still
think a flag to disable the whole thing would be less risky.

**Comment by mrjana** [comment]
I have to characterize the host FS mount because in my test I never remounted by rootfs to `mand` but when I remounted my `container.hosts` file to mand it somehow propogated to the rootfs as well. That may be required for this to work.

But we don't need `/var/lib/docker/containers' to be in tmpfs. It may just have to be remounted to carry the `mand` option so it's not that sketchy.

**Comment by thockin** [comment]
At least on my system, you can not remount a directory with `mand` unless
the root FS is also mounted `mand`.  Mounting tmpfs with `mand` DOES work,
though this means none of your state survives reboot (and is suddenly
orders of magnitude more expensive).

It's disappointing that this is still racy - if I already have the file
open()ed while docker tries to write it, it WILL get corrupted.  I just
wrote a bunch of tests that convinced me of it.  in process A: open, seek
to non-zero, sleep(5), write.  In process B: open, take lock, write,
sleep(10), exit.  Process B will write its contents, the lock will prevent
process A from write()ing until the lock is released, but eventually A will
write at whatever file offset it was at - corruption.

I still think the only complete answer is to nsenter the mount namespace of
every container, write a tmp file and rename() it to /etc/hosts.

More importantly: I think we're focusing on the race, but I am not
convinced that the race is what is causing the corruption - we don't write
to /etc/hosts in our tests.  I think there are bugs in the content-merging
logic that is forgetting to remove some lines and occasionally removing
lines it shouldn't.  I think THAT is far more important.

**Comment by rade** [comment]
> I still think the only complete answer is to nsenter the mount namespace of
every container, write a tmp file and rename() it to /etc/hosts.

Please, please, please do not do this. At least not without *also* providing a way to disable the entire `/etc/hosts` updating.

**Comment by brendandburns** [comment]
Yes, please provide the ability to disable `/etc/hosts` updating.  Users should be allowed to choose whether or not they want this feature.   Perhaps the right approach is a per-container option in the API?

**Comment by rade** [comment]
> per-container option in the API

+1

**Comment by thockin** [comment]
@rade why is that worse than what's happening today (or the proposed 90% fix to the race)?  At least the atomic replacement will not corrupt the file.

**Comment by rade** [comment]
it is worse because it breaks the existing hacks that prevent `/etc/hosts` updating.

**Comment by thockin** [comment]
@rade: oh yeah, excellent point.  @mrjana I agree 1000% with @rade now.  We're looking at ways to use volume mounts to mount our own /etc/hosts file and "hide" docker's (mis)management of this file completely.  At that point it's just a waste of cycles for docker to do it.

FWIW I still agree that we should be able to turn this off - this is a step to far for many people/systems.  That said, it sounds like Docker folks are VERY against giving users the ability to swap the batteries (to borrow an analogy) on this.  I am not sure why they are so adamant, and quietly so - all such requests have not been met with "no we won't do that because ..." they are just ignored.

I get that this sort of "poor man's service discovery" is part of Docker's plan to be everything to everyone, but it's currently totally broken, and it's not useful (or is actively harmful) for anyone using higher-level orchestration.

**Comment by icecrime** [comment]
> That said, it sounds like Docker folks are VERY against giving users the ability to swap the batteries (to borrow an analogy) on this. I am not sure why they are so adamant, and quietly so - all such requests have not been met with "no we won't do that because ...". They are just ignored.

I replied to your email regarding why I wouldn't do that. I surely didn't ignore your question, but you apparently chose to ignore my answer. I will just paste it again here:

>  I feel that making /etc/hosts content consistent for all containers running on a given host is the most important thing to fix right now. And yes, it does leave open the possibility of a container-side race condition on read, however I'm not confident going down the road of mandatory locking during a code freeze.
>
> Docker philosophy is to provide a consistent experience backed by swappable implementations. Docker 1.8 has been out for a couple months now, and users have come to expect that containers discover each other as they start: giving the option to disable it makes for very different user experiences depending on the daemon configuration.
>
> [...]
>
> From my point of view the feature being unreliable is an argument to fix it, not an argument for allowing to turn it off. It's unfortunate that 1.9.0 most likely won't ship with a definitive fix, but it will make the issue much less significant than it is today.

**Comment by rade** [comment]
> giving the option to disable it makes for very different user experiences depending on the daemon configuration.

The proposal by @brendandburns was to make this configurable *per container*.

**Comment by thockin** [comment]
@icecrime I don't think I am on an email thread with you - it must have been Brian.  If I am, I must have totally missed it, so I apologize.  Indeed, I just searched and I archived the thread accidentally.  My sincerest apologies.

**Comment by thockin** [comment]
Good news (for Kubernetes anyway) - I think we have a complete workaround.  I'd like to get someone from Docker (@icecrime?) to ACK that this is safe to do and will not get broken in the future.

Given how Kubernetes works, creating our "infrastructure container" before app containers, we can craft our own /etc/hosts file and use docker's volumes (-v outside:inside) to mount it into app containers.  I.e.

```
ID=$(docker run ... kubernetes/pause)
docker inspect $ID | grep IP
cat > /var/lib/kubelet/pods/$POD/hosts <<EOF
127.0.0.1 localhost
$IP $POD
EOF
docker run --net=container:$ID -v /var/lib/kubelet/pods/$POD/hosts:/etc/hosts ... user/appimage
```

This seems to work.  Docker will still manipulate /var/lib/docker/containers/*/hosts but we won't see that because we mounted over it.  Is this safe?  Can you commit to this continuing to work?

**Comment by thockin** [comment]
@artfulcoder pointed me at https://github.com/docker/docker/pull/14613 in which this feature was explicitly enabled.  Can I safely assume this will not break in the upcoming changes in this area?

**Comment by estesp** [comment]
Yes, we (one of our IBM security researchers) added #14613 and then added read-only capability in #14965.  Given our public container service depends on this capability (for different reasons than yours, albeit) I think we can safely assume this is a workable/supported solution.  However, @icecrime or the libnetwork team can confirm as I'm not aware of anything that will "break" this coming around the corner?

**Comment by cpuguy83** [comment]
I'm slightly surprised that docker is not manipulating the mounted hosts files, especially since the path to the mounted file is stored in the container struct, but also ok with it.

**Comment by cpuguy83** [comment]
Since this is the case, and it's already in 1.8 this way too ( :smile: ) I'd recommend that we have libnetwork not even try to update the `/var/lib/docker/containers/<id>/hosts` file too.

**Comment by mrjana** [comment]
> It's disappointing that this is still racy - if I already have the file
open()ed while docker tries to write it, it WILL get corrupted.  I just
wrote a bunch of tests that convinced me of it.  in process A: open, seek
to non-zero, sleep(5), write.  In process B: open, take lock, write,
sleep(10), exit.  Process B will write its contents, the lock will prevent
process A from write()ing until the lock is released, but eventually A will
write at whatever file offset it was at - corruption.

The POSIX locking semantics works on file region level. Not on the whole file level. So obviously when a file is open for writing but you haven't written anything there is no way for the kernel to reject your write lock on a region (which most commonly you do on the whole file but it is just a special case for the kernel). And if you are determined to write at that location it is going to be a problem. This is the sorry state of affairs with file locking in linux

**Comment by mrjana** [comment]
@thockin 
> I still think the only complete answer is to nsenter the mount namespace of
every container, write a tmp file and rename() it to /etc/hosts.

This will still not solve the problem because `rename()` provides atomicity only if you are just trashing the old content and just writing something totally new. If you have to read/modify/write the file which is what you do most of the time because you don't want to wipe out whatever the container app has written in the /etc/hosts, then this is also not a good solution. In fact it will be more damaging to wipe out the existing  contents of /etc/hosts. If nobody is going to update the /etc/hosts inside the container then there is no issue with racy behavior to begin with

**Comment by mrjana** [comment]
@thockin 
> More importantly: I think we're focusing on the race, but I am not
convinced that the race is what is causing the corruption - we don't write
to /etc/hosts in our tests.  I think there are bugs in the content-merging
logic that is forgetting to remove some lines and occasionally removing
lines it shouldn't.  I think THAT is far more important.

The problem you are facing in 1.8.3 is not because of some erroneous content merging logic. In fact we don't do a lot of complicated content merging logic. What's happening is Docker container start/stop code can run many go routines and they potentially can step over each other writing to the /etc/hosts file of a single container and that can result in one goroutine trying to delete one line by reading the file, deleting a line and writing back the file, while another goroutine doing the same but whatever it deleted has been put back by the first go routine. We have simply solved this problem for now by synchronizing within docker itself. See this PR for reference:

https://github.com/docker/libnetwork/pull/689

**Comment by brendandburns** [comment]
@mrjana 
> If nobody is going to update the /etc/hosts inside the container then there is no issue with racy behavior > to begin with

actually, there is racy behavior even if the container doesn't modify the contents of `/etc/hosts`.  The trouble is that the existing behavior uses truncate and write to re-write the entire file.  So a reader can catch the file in mid write and either get an empty, incomplete, or even corrupted file.

**Comment by thockin** [comment]
If you need to do RMW inside the namespace you could use F_SETLEASE to
block readers (but you can't have bind-mounted it).  It still requires exec
on every container every time a container starts or stops, which is really
heavy.

As said elsewhere, those of us who are (or soon will be :) bind mounting
our own /etc/hosts DO NOT want docker trashing it, so I don't think you
should do this anyway.

Brendan is right, though - anything you do that is not either atomic or
blocks readers WILL cause random glitching for apps in containers.  It's a
pretty bad situation for a "production ready" system, IMO.

**Comment by mrjana** [comment]
If we want to block readers we could still do mandatory locking which works across bind mounts and will protect readers from getting random stuff especially since we have to do RMW and `rename` doesn't work across bind mounts.

But one more thing I would like to try is bind mount the directory containing the hosts file and not the hosts file itself and then try doing `F_SETLEASE`. Because if I understand kernel code correct in that case the `dentry` of the file will not be greater than one and so `F_SETLEASE` will not fail with `EAGAIN`

**Comment by thockin** [comment]
Yeah if you mount a whole directory it should work, but that directory is
/etc, which comes from the container image.  You could do something like
mount onto /var/lib/docker-managed and then symlink /etc/hosts ->
/var/lib/docker-managed/hosts.  Blech.

**Comment by ArtfulCoder** [comment]
The symlinking should be done only if the user has NOT already mounted on /etc/hosts, so that #14613 continues to be supported.

**Comment by thockin** [comment]
yeah, frankly most proposed solutions are pretty hacky and way too delicate
to be considered really.

**Comment by mrjana** [comment]
@thockin What is the problem with bind mount into different directory and symlink inside container? Why is that a hack if it solves the race 100%?

**Comment by thockin** [comment]
@mrjana just because it's so delicate and unexpected.  You have to NOT do it if I bind mount my own /etc/hosts.  You have to remove the file from my image (causing a write in my writable layer).  You have to not do it if I am `--net=host`.  You have to claim a new space to mount the `docker-managed` dir.  IMO it's just too brittle for too little value.

**Comment by tianon** [comment]
I'm definitely very surprised with recent Docker builds to see all my containers showing up in /etc/hosts.  Having links control this type of information sharing in an explicit way is a nice feature, but I honestly see this level of unfettered exposure as a serious bug, and hope that we adjust this behavior to be opt-in and disabled by default.

**Comment by thockin** [comment]
@tianon one man's bug is another man's feature.  It sounds like this is very much intentional and there's no chance of convincing people this is a bad idea.

**Comment by tianon** [comment]
Yeah, I have no problem with the feature, but it being on by default seems like a really grave bug IMO.

**Comment by vieux** [comment]
I agree with @tianon it's a big bug IMO

```
 docker run -d --name test busybox ping google.com
97a0be3eed7ba31aaa686d1f8893f7470e89936a26b018381c63877bd83c6d0d
$ docker run  --name test2 busybox cat /etc/hosts
172.17.0.3	228d3aaf47f9
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
172.17.0.2	test
172.17.0.2	test.bridge
```

1) Why is the name of container `test` leaking in `test2` it doesn't make sense.
2) Why `172.17.0.3` is only for the container id, what about it's name `test2`

I think this "feature" with big " should be removed from 1.9.0

**Comment by tianon** [comment]
This would be even more interesting if you had a container named `google.com` before running that ping.

**Comment by mrjana** [comment]
@vieux I don't understand. Apart from the discussion from wanting a flag to disable automatic container linking, this is what automatic container linking is. i.e since `test` and `test2` is part of the same `bridge` network they will both learn each other's name.

**Comment by mrjana** [comment]
@vieux 2) was always the case even before 1.8.3. When you create a container without linking it's container id and it's IP address will be present in /etc/hosts

**Comment by mrjana** [comment]
@tianon That is kind of an argument like saying that I will call my host `google.com` but when I ping `google.com` I want to reach the real `google.com` not localhost which is also named as `google.com`. We can't protect users if they are more than willingly shooting themselves in the foot because they have a gun.

**Comment by ArtfulCoder** [comment]
@mrjana
1) I think what @tianon meant is containerA has an entry "google.com" in its hosts file because another containerB came up behind the same bridge with the name "google.com", which containerA did not expect, and potentially, containerA does not control or is aware of the existence of containerB.
This can lead to Denial of Service attacks.

2) Also, if containerB with the name "google.com" goes away, dns clients on containerA will suddenly start resolving to the real google.com since the corresponding containerB entry in the hosts file of containerA will have gone..

**Comment by vieux** [comment]
@mrjana I understand why it's happening , it makes sense. but it's a big change compared to before

**Comment by mrjana** [comment]
@ArtfulCoder If you are thinking about a security issue here(I am assuming that is what you meant here since you mentioned "DoS attacks"), this feature does not enhance and reduce the security of the system in any which way. Somebody has to have enough privileges to create a container in a host. If that somebody has malicious intent, then they can as well go and edit the /etc/hosts file for the container to add an entry for `google.com` and cause the same effect.

Also there is nothing very different here than what happens in the case of a real DNS. If you added a DNS record with `google.com` in your local DNS servers you will resolve to that IP with DNS clients internal to your network pointing to that DNS server. If you remove the entry, then the clients will resolve to the real `google.com`. What's the difference here?

**Comment by ArtfulCoder** [comment]
I was thinking in terms of a multi-tenant scenario, but maybe that is not a
use case docker naturally supports if containers are behind the same bridge
?

**Comment by thockin** [comment]
Note that what gets written to this file is somewhat sanitized, so a
container named "google.com" becomes "google-com".

**Comment by ArtfulCoder** [comment]
so, it can could collide with a container actually named "google-com" ?

**Comment by thockin** [comment]
pfft, details :)

**Comment by ibuildthecloud** [comment]
oh shoot... I never really noticed this.  I'm don't think its a good idea to always edit the hosts files for all containers in a network.  There has to be some way to opt in.  I was fine with docker links doing this because it's opt in.  This was not the behavior in 1.7.  In Rancher we always name containers as UUIDs so I never noticed this, but we were just about to put in a patch to put the "friendly" name on the container but that is going to mess with DNS based service discovery since /etc/hosts takes priority.  I don't want container names accidentally taking precedence if somebody is using something like consul or skydns.

Sidenote, one really odd thing is that you can discover the names of all containers in the network, but not your own name...

I think this should be an option on network create and the default `bridge/docker0` network should have it off by default.  Then things like compose can create a network and turn this on as it seems that compose will rely on this behavior now.

**Comment by thockin** [comment]
Also, this allows other non-valid-for-DNS characters into the /etc/hosts file.

**Comment by crosbymichael** [comment]
Ok, so lets do this.  Fix the bug/regression that caused all containers to be written to the etc/hosts file for containers running on our default bridging setup.  I'm sure @mrjana or someone will take care of this for us and after the PR is open make sure its linked here to resolve the issue.

Ez bug, ez fix.

![just do it](http://www.theloop.ca/wp-content/uploads/2015/06/mansw.gif)

**Comment by thockin** [comment]
Is this s really closed?  Assuming the corruption is fixed, it is still a
racy implementation for any non-default networks, which is sort of the hot
new thing to use isn't it?

I.e. the upcoming swarm release with overlay network will be flake-tastic.

**Comment by mfischer-zd** [comment]
I agree with @thockin: This issue is still open, as far as I can tell.

It seems to me that there are at least three issues regarding /etc/hosts generation:

1. Race conditions involving its manipulation
2. Link entries being added, whether or not `--link` was specified in the `docker run` invocation
3. Entry not being updated to new IP if linked container is restarted (maybe related to (1)?)

So far it looks like only (2) was fixed by updating libnetwork.  The other two issues remain, as far as I can tell.

If you'd like to close the plethora of open issues related to /etc/hosts in favor of another tracking bug, that's fine, but I think it should be clear what that tracking issue is.

**Comment by mfischer-zd** [comment]
I agree with @thockin: This issue is still open, as far as I can tell.

It seems to me that there are at least three issues regarding /etc/hosts generation:

1. Race conditions involving its manipulation
2. Link entries being added, whether or not `--link` was specified in the `docker run` invocation
3. Entry not being updated to new IP if linked container is restarted (maybe related to (1)?)

So far it looks like only (2) was fixed by updating libnetwork.  The other two issues remain, as far as I can tell.

If you'd like to close the plethora of open issues related to /etc/hosts in favor of another tracking bug, that's fine, but I think it should be clear what that tracking issue is.

**Comment by mavenugo** [comment]
@mfischer-zd I hope you are not expecting to use this issue to track all the /etc/hosts related issues (including the link specific ones). This issue was opened in 1.8 for the default bridge where the container /etc/hosts were updated with names even when links were not used. With #17325 that is resolved.

I think the right answer is https://github.com/docker/docker/issues/17195 and am supportive of that.

**Comment by mfischer-zd** [comment]
@mavenugo I agree with the direction, but I do not think this issue can be closed until #17195 has actually been implemented.

**Comment by mavenugo** [comment]
@mfischer-zd why is that ? this issue was filed for 1.8 (which means the default bridge network) and with #17325 that behavior is restored. Hence, technically this issue is resolved. Unless we want to use this issue to track the networks beyond the default bridge network, which btw is not available in 1.8.
Hence it is more appropriate to put effort on #17195 and work together to get it implemented ASAP.

**Comment by thockin** [comment]
Will #17195 land in time for 1.9?  If not, the raciness issue still exists,
at least.  It is mitigated by being only non-default networks, but isn't
that a huge selling point of 1.9?  I would expect a lot of people to try it
out, and flakes are the worst problems to debug...

**Comment by mavenugo** [comment]
@thockin I hope we all understand release-cycles and issue management :) My responses to @mfischer-zd is to the technical aspect of issue management and the fact that this issue was filed on 1.8.3 and is applicable to the default bridge network. And hence technically #17325 resolved this. Wont you agree to that ?
And with https://github.com/docker/docker/issues/17278 we have resolved the internal race issues as well.

With a healthy upstream/downstream relationship with Kubernetes and Docker, I hope this time around, you would be able to try out RCs early in the release cycle and not wait for x.x.3 release to give feedback.

**Comment by mfischer-zd** [comment]
Ultimately I leave the decision to you, @mavenugo - my only concern is that /etc/hosts management is in a bit of a quagmire at the moment (though it looks like you're making good progress on fixing them) and there is a large number of issues to sift through to figure out just which issue affects a particular user.  

I don't mean to suggest that this is the appropriate issue to track those with, though.  What I was trying to say is:

* This issue is about race conditions in /etc/hosts management
* Some issues around it have been addressed already
* You claim the fix is in #17195 - but that has not been completed
* You also claim the fix is in #17325 - but unless that fixes all race conditions that could causes /etc/hosts corruption, I respectfully disagree.

**Comment by thockin** [comment]
Until there's an open issue specifically about not corrupting reads of
/etc/hosts I think we DON'T all understand issue management.  It's
attractive to close this one out, since it got some spotlights on it, but
let's not pretend it's entirely fixed for 1.9.

#17278 sounds like it fixes the biggest problem, though my casual reading
suggests that a hosts file for a shared net namespace (--net=container)
will be processed more than once.

As for zinging us on filing bugs late - I said it before, this bug took is
a LONG time to pin down (though it wasn't the first related bug I filed).
We will do everything we can to file bugs early and often, but we have our
own software to debug, too.

**Comment by mrjana** [comment]
@thockin  As I said before, the racy behavior existed since `--link` existed so let's not think that this racy behavior was introduced in 1.9. I agree it's not fixed in 1.9 but it's not fixed in 1.8, 1.7, 1.6 or 1.5 and so on as well. It just became more common in 1.8.3 because docker itself was contributing for the race and was reported by you via this issue and we fixed it. Since this issue was reported for that specific problem it is appropriate to close the issue. I am perfectly fine to track the racy behavior issue as a separate issue. But let's not mix a lot of things and confuse users.

Also @mfischer-zd now that the default bridge network does not do automatic service discovery please retest to see if issue number (about /etc/hosts not getting updated if a linked container is restarted) is still seen. If seen it would be helpful to file a separate issue for that because it is definitely not the same as racy behavior issue.

**Comment by mrjana** [comment]
@thockin 
> #17278 sounds like it fixes the biggest problem, though my casual reading
suggests that a hosts file for a shared net namespace (--net=container)
will be processed more than once.

A shared net namespace (`--net=container`) will not even invoke libnetwork for the child containers. Once the parent container is setup that is all there is to it. All the children simply bind mount the the hosts file created for parent container. So I don't know what in that #17278 gave you the impression that it is getting processed more than once in the shared net namespace case.

**Comment by thockin** [comment]
> @thockin As I said before, the racy behavior existed since --link existed so let's not think that this racy behavior was introduced in 1.9. I agree it's not fixed in 1.9 but it's not fixed in 1.8, 1.7, 1.6 or 1.5 and so on as well. It just became more common in 1.8.3 because docker itself was contributing for the race and was reported by you via this issue and we fixed it. Since this issue was reported for that specific problem it is appropriate to close the issue. I am perfectly fine to track the racy behavior issue as a separate issue. But let's not mix a lot of things and confuse users.

All I am saying is there isn't a bug that I know of that tracks the
fact that in-place editing /etc/hosts is going to cause user problems.
And let's not be disingenuous - yes it existed forever, but with 1.9
you're encouraging people to create Network objects and as soon as
they do, Docker is back to "contributing for the race".

Truth be told, it doesn't matter for us - we have a workaround.  I am
just advocating that we make sure this is an understood shortcoming of
networks in 1.9

**Comment by thockin** [comment]
Maybe I misunderstood then.  I understood that the corruption was parallel
goroutines accessing the file and I assumed that was once per
namespace-user.  If not that, then what parallel goroutines?  Just multiple
starts/stops in parallel?

Thanks, sorry for any FUD - honest misunderstanding.

**Comment by mrjana** [comment]
> And let's not be disingenuous - yes it existed forever, but with 1.9
you're encouraging people to create Network objects and as soon as
they do, Docker is back to "contributing for the race".

I am not sure which race you are talking about. But if you are talking about Docker contributing to the race (oh I just corrected my grammatical mistake that I made when I typed by message in a hurry and I am so glad how you pointed out my mistake so subtly by repeating the same mistake in quotes. That was very kind of you) by even writing to /etc/hosts file then you don't have to create Network objects and you don't need 1.9. You just need to use links in any previous release it is supported. I don't know where I am being disingenuous here.

If you are talking about Docker contributing to the race by having multiple threads of its own trying to write to the same file, then that is fixed whether or not you use 1.9's new Network objects.

**Comment by mrjana** [comment]
> Maybe I misunderstood then.  I understood that the corruption was parallel
goroutines accessing the file and I assumed that was once per
namespace-user.  If not that, then what parallel goroutines?  Just multiple
starts/stops in parallel?

Yep, just single start running in parallel with a lot of stops.

**Comment by thockin** [comment]
On Sun, Oct 25, 2015 at 10:24 PM, Jana Radhakrishnan
<notifications@github.com> wrote:

> Yep, just single start running in parallel with a lot of stops.

Got it, thanks.  Sorry for the confusion.

**Comment by thockin** [comment]
On Sun, Oct 25, 2015 at 10:23 PM, Jana Radhakrishnan
<notifications@github.com> wrote:
>
> And let's not be disingenuous - yes it existed forever, but with 1.9
> you're encouraging people to create Network objects and as soon as
> they do, Docker is back to "contributing for the race".
>
> I am not sure which race you are talking about. But if you are talking about Docker
> contributing to the race (oh I just corrected my grammatical mistake that I made when
> I typed by message in a hurry and I am so glad how you pointed out my mistake so
> subtly by repeating the same mistake in quotes. That was very kind of you) by even

Sorry, wasn't meant to be a jab, I just thought it was more context.
It came across badly.

> writing to /etc/hosts file then you don't have to create Network objects and you don't
> need 1.9. You just need to use links in any previous release it is supported. I don't know
> where I am being disingenuous here.

One of the big features of 1.9 (as I understand the plan) is networks.
Using networks brings up the live-editing-causes-read-corruption
problem which is NOT solved.  I expect that using networks, the
live-edit operation will happen far more frequently than with links.
Ergo, users will experience problems at a far higher rate.

Yes, "don't use networks" is a valid workaround, but it's not one I
expect you to advocate too strongly for :)

The disingenuity is in closing this bug without filing another issue
that details the still-existing read-corruption in 1.9.x.  That's all
I am agitating for.

**Comment by mrjana** [comment]
> The disingenuity is in closing this bug without filing another issue
that details the still-existing read-corruption in 1.9.x.  That's all
I am agitating for.

Oh, we don't mean to hide it under the carpet. We will properly document it in the Caveats section of 1.9 release(this is all already being planned for and I am not making this up on the fly) and we will file a proper issue describing the problem in greater detail so everybody is aware. This bug was closed by automatic github workflow because we marked this issue as "Fixes" in #17325. Also it is not going to be an effective way to use this issue to track and educate users with the racy behavior issue. It may confuse the users than providing them clarity. That is why we need a separate issue.

You can rest assured that we are not dropping the ball on this.

**Comment by thockin** [comment]
Sounds fair to me.  Thanks.

**Comment by aidanhs** [comment]
Hang on, am I to understand that you're removing a feature people are relying on (in my case, relying on it extensively) between rc2 and the official release?

I can see that having this blindly enabled all the time is problematic and I'm definitely in favour of having opt-out, but it looks like there's a workaround upthread and I've been relying on this behaviour for two months now, since 1.8.0 came out. My impression from https://github.com/docker/docker/issues/16619#issuecomment-143641789 was that this would just be a flag to disable the behaviour (even if it's on the daemon level).

Make no mistake, this is a breaking change and will cause significant misery for a number of users - the feature is an incredible convenience for a number of reasons, and in some cases it enables use-cases that would otherwise be impossible (for example, exposing an [SSH agent](http://aidanhs.com/blog/post/2015-10-07-dockerfiles-reproducibility-trickery/#_streamlining_your_experience_using_an_ssh_agent) inside docker build).

**Comment by mfischer-zd** [comment]
@aidanhs I've been looking at the documentation and release notes for Docker and I can't seem to find any mention of the (now reverted) behavior in question being documented anywhere.  

Bugs and mistakes happen; relying on undocumented behavior is going to cause problems for you.

As far as I can tell, the only documented way to get Docker to automatically allow containers to address each other is to specify the `--link` option to `docker run`.  If this changes, it needs to be published; only then can one reasonably rely on it (at least until the next major revision is released).

**Comment by aidanhs** [comment]
@mfischer-zd I'd see this as equivalent to a new behaviour being added to the VOLUME command in dockerfiles such that if you do `VOLUME /etc:/outeretc`, it mounts `/etc/ from the host into the container.

That functionality would be fairly nasty, but if it had lasted 2.5 months, spanning 3 official releases and two release candidates, I'd understand if the people who *did* like that functionality at least wanted it available as an argument to `docker build` given that they may well have built some kind of process around it.

There are 4 different ways the functionality could be made available (excluding reverting the revert) listed at https://github.com/docker/docker/pull/17325#issuecomment-151233750, but none are available right now.

I'd understand if this was a minor convenience or clearly a hack, but saying 'tough luck' about the removal a feature which feels intentional and seems to align with "batteries included" is fairly uncharitable.
